src/Core/Content/MailTemplate/Service/MailTemplateService.php

```diff
@@ -0,0 +1,241 @@
+                $exception->getTemplateLine(),
+            )];
+        }
+
+        return $this->validateVariables($usedVariables, $availableVariables);
+    }
+
+    /**
+     * @param array<string, string> $usedVariables
+     *
+     * @return MailTemplateValidationResponse[]
+     */
+    public function validateVariables(array $usedVariables, EventDataCollection $availableVariables): array
+    {
+        $errors = [];
+
+        foreach ($usedVariables as $var) {
+            if ($this->isKnownVariable($var, $availableVariables, $errors)) {
+                continue;
+            }
+
+            $this->validateComplexVariable($var, $availableVariables, $errors);
+        }
+
+        return $errors;
+    }
+
+    /**
+     * @param MailTemplateValidationResponse[] $errors
+     */
+    private function isKnownVariable(string $var, EventDataCollection $availableVariables, array &$errors): bool
+    {
+        if ($field = $availableVariables->get($var)) {
+            if (!($field instanceof ScalarValueType)) {
+                $errors[] = new MailTemplateValidationWarning(
+                    $this->dataValidator,
+                    MailTemplateValidationWarning::TYPE_COMPLEX_ELEMENT,
+                    ['variable' => $var]
+                );
+            }
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param MailTemplateValidationResponse[] $errors
+     */
+    private function validateComplexVariable(string $var, EventDataCollection $availableVariables, array &$errors): void
+    {
+        if (!str_contains($var, '.')) {
+            $errors[] = new MailTemplateValidationError(
+                $this->dataValidator,
+                MailTemplateValidationError::TYPE_UNKNOWN_VARIABLE,
+                ['variable' => $var]
+            );
+            return;
+        }
+
+        $varParts = explode('.', $var);
+
+        $nestedAvVars = $availableVariables->get($varParts[0]);
+
+        for ($i = 0; $i < \count($varParts); ++$i) {
+            if (!$nestedAvVars) {
+                $errors[] = new MailTemplateValidationError(
+                    $this->dataValidator,
+                    MailTemplateValidationError::TYPE_UNKNOWN_VARIABLE,
+                    ['variable' => $var]
+                );
+                break;
+            }
+
+            if ($nestedAvVars instanceof ScalarValueType && $i === \count($varParts) - 1) {
+                break;
+            }
+
+            if ($nestedAvVars instanceof EntityType) {
+                $prefix = $varParts[$i];
+                for ($j = 0; $j <= $i; ++$j) {
+                    unset($varParts[$j]);
+                }
+
+                $path = \implode('.', $varParts);
+
+                $errors = $this->validateEntityField($path, $prefix, $nestedAvVars->getDefinitionClass(), $errors);
+
+                break;
+            }
+
+            if ($nestedAvVars instanceof ArrayType) {
```

Comment:
same here. don't use fancy abbreviations. `nestedAvVars` is `NestedAvailableVariables`